### PR TITLE
[SILVerifier] Add detection of conflicting debug variables

### PR DESCRIPTION
**Pull request by Snowy1803**
IRGenDebugInfo assumes, using a cache, that all variables with the same scope, same name and same source location, are lowered to the same type. This assumption was broken in some rare cases. This PR fixes those cases and adds a SILVerifier assertion to detect them early.

rdar://124555787

### CODE REVIEW

**Comment by Snowy1803** [comment]
@swift-ci please test source compatibility

**Comment by Snowy1803** [comment]
@swift-ci please test

**Comment by Snowy1803** [comment]
@swift-ci please test source compatibility

**Comment by adrian-prantl** [comment]
looks like `test/SILOptimizer/init_accessors.swift` is triggering the check

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 1]
```diff
@@ -792,6 +792,32 @@ static void checkAddressWalkerCanVisitAllTransitiveUses(SILValue address) {
   llvm::report_fatal_error("invoking standard assertion failure");
 }
 
+} // end anonymous namespace
+
+namespace llvm {
+
+template<>
+struct DenseMapInfo<SILLocation> {
```
Comment (adrian-prantl) [inline lib/SIL/Verifier/SILVerifier.cpp:800]: Can you move this to SILLocation.h?

**Comment by Snowy1803** [comment]
@swift-ci please smoke test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
@swift-ci test linux

**Review by adrian-prantl — APPROVED** [review]

**Pull request merged**